commands/src/main/java/discord4j/commands/CommandBootstrapper.java

```diff
@@ -0,0 +1,163 @@
+     *
+     * @see #attach(discord4j.core.DiscordClient)
+     * @param providers The {@link discord4j.commands.CommandProvider}s to initialize this instance with. Note that
+     * this set is expected to be thread-safe as there will be concurrent read access required.
+     */
+    public CommandBootstrapper(Set<CommandProvider> providers) {
+        this.providers = providers;
+    }
+
+    /**
+     * Constructs the message listener for commands.
+     *
+     * @see #attach(discord4j.core.DiscordClient)
+     */
+    public CommandBootstrapper() {
+        this(new CopyOnWriteArraySet<>()); //TODO: LinkedHashSet is probably better, but not concurrent!
+    }
+
+    /**
+     * Attaches this {@link CommandBootstrapper} instance to the
+     * {@link discord4j.core.event.domain.message.MessageCreateEvent} of the passed in client.
+     *
+     * @return A flux (that need not be subscribed to), which signals the completion of command executions.
+     */
+    public Flux<? extends BaseCommand> attach(DiscordClient client) {
+        return client.getEventDispatcher()
+                     .on(MessageCreateEvent.class)
+                     .flatMap(event -> dispatcher.dispatch(event, providers, errorHandler))
+                     .share();
+    }
+
+    /**
+     * Replaces the {@link CommandDispatcher} used by this instance. This can be used to implement
+     * custom command discovery rules.
+     *
+     * @param dispatcher The new dispatcher to use.
+     * @return The same {@link CommandBootstrapper} instance to chain.
+     */
+    public CommandBootstrapper setCommandDispatcher(CommandDispatcher dispatcher) {
+        this.dispatcher = dispatcher;
+        return this;
+    }
+
+    /**
+     * Replaces the {@link CommandErrorHandler} used by this instance. This is used to
+     * implement unified error handling despite command implementations.
+     *
+     * @param handler The new error handler to use.
+     * @return The same {@link CommandBootstrapper} instance to chain.
+     */
+    public CommandBootstrapper setCommandErrorHandler(CommandErrorHandler handler) {
+        this.errorHandler = handler;
+        return this;
+    }
+
+    /**
+     * Gets the {@link CommandDispatcher} being currently used.
+     *
+     * @return The current {@link CommandDispatcher}.
+     */
+    public CommandDispatcher getCommandDispatcher() {
+        return dispatcher;
+    }
+
+    /**
+     * Gets the {@link CommandErrorHandler} being currently used.
+     *
+     * @return The current {@link CommandErrorHandler}.
+     */
+    public CommandErrorHandler getErrorHandler() {
+        return errorHandler;
+    }
+
+    /**
+     * Gets the command providers registered with this dispatcher.
+     *
+     * @return The providers. It is expected that this returns an immutable copy of the internal backing set.
+     */
+    public Set<CommandProvider> getCommandProviders() {
+        return Collections.unmodifiableSet(providers);
```

Comment:
I don't see the point when we're allowing mutation of the Set anyway. Since it's safe to use concurrently, there should be no need to disallow raw access.

Reply:
I'd rather we disallow modification to the raw set. Allowing it may lead to unintended side effects for users and it removes control from us which may be useful in the future if we want to add some kind of hooks.